### DISCUSSION TITLE
Status of EMap usage with EMF.cloud technologies

### DISCUSSION THREAD

**Discussion by vhemery** [eclipse-emfcloud/emfcloud#155] [ModelServer] [answered]
Hello,
I am having troubles trying to work with an EMap reference (a multivalued reference with type pointing to an EStringToStringMapEntry Eclass, with `java.util.Map$Entry` instance type name).
Has this been tested before with EMF.cloud or am I going into unexplored territory ?
I haven't found a trace of the EAnnotation entries in the Theia Ecore tools... (which is basically the same as my use case)

Thanks for your feedback.

So far, here are the situations that I have encountered :
- command to create a new entry impacts the model correctly. 👍 
- this new entry has a `null` key value, which makes `org.eclipse.emfcloud.jackson.databind.ser.EMapStringSerializer.serialize(EList<Entry<String, ?>>, JsonGenerator, SerializerProvider)` crash with NPE. It looks like EMapStringSerializer could be refactored into something more generic and make use of some jackson options like mentioned on https://www.baeldung.com/jackson-map-null-values-or-null-key (we could try using com.fasterxml.jackson.databind.ser.std.MapSerializer, but a simple fix to replace null by `""` seems to work for my limited tests)
- even if it passes, it looks like the json data is not consistent with the schema generated by `org.eclipse.emfcloud.modelserver.jsonschema.DefaultJsonSchemaConverter.createJsonSchema(EReference, boolean)`. The schema is generated with an array of entries, whereas the Jackson serializer generates a Map object (keys and values). So when trying to use JsonForms, `isObjectArrayWithNesting` returns true, but `data` constructed by `withJsonFormsArrayLayoutProps` (delegates to `mapStateToArrayLayoutProps`) is `props.data.length` which is undefined because `props.data` is an object.

For this latest situation, I think it would be better to fix the schema and keep the Map object (not an array). Would you second that ?

**Comment by eneufeld** [comment, marked as answer]
Hi @vhemery , 

sorry for the late reply.
We checked internally and nobody used EMF.cloud with EMaps until now.
As you already figured out, there is only rudimentary support.
We are happy to support you if you need to add support for EMaps.

**Reply by eneufeld** [reply]
I agree that a map should be mapped to an object and not an array. So the schema should be fixed accordingly.
